content/posts/marimo.md

```diff
@@ -0,0 +1,253 @@
+Thanks to Pyodide, you can use any pure Python package, as well as a number of
+packages for scientific computing, including Pandas, NumPy, SciPy, matplotlib,
+and scikit-learn, entirely in the sandbox of your web browser --- _without
+installing Python on your machine and without paying for infrastructure_.
+
+## Implementation: Porting marimo to Pyodide
+
+Porting marimo to work with Pyodide was relatively straightforward, a testament to the
+latter's robustness and thorough documentation.
+
+marimo's original implementation has three main parts:
+
+1. a Python codebase implementing a kernel that runs Python code;
+2. a TypeScript codebase that issues control commands for the Python kernel;
+3. a server that connects the TypeScript and Python codebases.
+
+All that was required were minimal changes to our TypeScript and Python
+codebases, and replacing the server with a lightweight bridge.
+
+### TypeScript
+
+marimo uses the same TypeScript build when running with Pyodide or CPython.
+This helps maintain consistency between the two experiences and avoid feature
+drift. The code path splits based on feature flags; when running under Pyodide,
+we download the marimo wheel from PyPI and initialize a single-threaded,
+Pyodide-compatible marimo kernel. Passing only a filename and a message
+callback function, we initialize the kernel as an async, never-ending Python
+process.
+
+We interact with this kernel through a lightweight RPC bridge that sends requests to
+the kernel and receives responses through the callback passed. This allows for
+an asynchronous-like feel to the execution. To support interruption of Python
+code, we simply followed the [Pyodide
+documentation](https://pyodide.org/en/stable/usage/keyboard-interrupts.html) on
+using its `SharedArrayBuffer`-based mechanism.
+
+For performance, we cache our assets and [run Pyodide in a web
+worker](https://pyodide.org/en/stable/usage/webworker.html) to avoid blocking
+the main browser thread. This creates some extra complexity with type-safety
+and mimicking blocking RPCs.
+
+We heavily leverage the Emscripten filesystem. Code is passed from the URL hash
+(for ease of sharing), to the main frame, to the worker, and finally to the
+Emscripten filesystem. Emscripten supports many filesystem implementations such
+as IndexDB which we use to persist the user's files. We are working on our own
+implementation of the Emscripten filesystem built on top of any S3-compatible
+bucket. This means you'll be able to list, read, and write files from your S3
+bucket just by interacting with ”os” filesystem all in Python.
+
+In order to make the experience smooth in any browser environment, we do our
+best to auto-install packages. Whenever a cell is run, we try to install it
+through Pyodide or otherwise fallback to micropip.
+
+### Python
+
+Similar to our TypeScript codebase, our Python codebase uses the same
+implementation for Pyodide and CPython. We introduced just a few branches based
```

Comment:
One more nit: Pyodide is CPython. I usually say "native Python" to mean "Python for mac/linux/windows" but it is a bit awkward.

```suggestion
implementation for Pyodide and native Python. We introduced just a few branches based
```